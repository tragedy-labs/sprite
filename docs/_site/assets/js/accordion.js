/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
