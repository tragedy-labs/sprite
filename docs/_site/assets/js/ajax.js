// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
