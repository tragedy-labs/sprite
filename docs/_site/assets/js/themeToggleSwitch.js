function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
