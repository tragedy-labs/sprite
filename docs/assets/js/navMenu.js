document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
