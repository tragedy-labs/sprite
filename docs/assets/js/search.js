import data from "/assets/js/searchData.js";
import serializedIndex from "./assets/js/serialized.js";
//import lunr from "lunr"; // assuming lunr is a separate module

// Constants
const MIN_SEARCH_LENGTH = 3;

// State definition
const state = {
  filters: {
    class: true,
    method: true,
    type: true,
  },
  results: {
    filtered: [],
    full: [],
  },
  visible: false,
};

// Elements for reference
const elements = {
  input: null,
  resultsBox: null,
  resultList: null,
  numberOfResults: null,
  numberResultsShowing: null,
  backdrop: null,
  filters: null,
  searchLabel: null,
  close: null,
};

// Initialize search
async function initializeSearch() {
  const root = document.getElementById("search_button");
  elements.input = document.getElementById("searchInput");
  elements.resultsBox = document.getElementById("searchResults");
  elements.resultList = document.getElementById("resultList");
  elements.numberOfResults = document.getElementById("numberOfResults");
  elements.numberResultsShowing = document.getElementById("numberResultsShowing");
  elements.filters = document.getElementById("searchFilters")
    .getElementsByTagName("input");
  elements.searchLabel = document.getElementById("searchLabel");
  elements.close = elements.resultsBox.querySelector("#close_search_dialog");

  // Add event listeners
  elements.input.addEventListener("input", search);
  elements.close.addEventListener("click", hideSearchBox);

  const openSearch = document.getElementById('search_button');

  openSearch.addEventListener('click', showResultsBox);
  document.addEventListener('keydown', (e)=>{
    if (e.key === '/') {
      e.preventDefault();
      showResultsBox();
    };
  }, false);

  // Load the index
  const index = lunr.Index.load(serializedIndex);

  // Initialize search filter toggles
  for (let i = 0; i < elements.filters.length; ++i) {
    elements.filters[i].addEventListener("change", () => {
      state.filters[elements.filters[i].value] = elements.filters[i].checked;
      updateResultList();
    });
  }

  // Define functions
  function appendBackdrop() {
    elements.backdrop = document.createElement("div");
    elements.backdrop.id = "backdrop";
    elements.backdrop.addEventListener("click", hideSearchBox);
    document.body.appendChild(elements.backdrop);
  }

  function removeBackdrop() {
    elements.backdrop.removeEventListener("click", appendBackdrop);
    elements.backdrop.remove();
    elements.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = elements.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === elements.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              elements.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection(e);
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    elements.resultsBox.classList.remove("hidden");
    elements.input.focus();
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    elements.resultsBox.classList.add("hidden");
    document.removeEventListener("keydown", checkKeyPress);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    elements.resultList.innerHTML = "";
    // elements.numberOfResults.innerHTML = state.results.full.length;
    // elements.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");
          li.classList.add(doc.type);

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          elements.resultList.appendChild(li);
        }
      });

      //elements.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  function search() {
    if (elements.input.value.length > 0) {
      elements.searchLabel.classList.add("hidden");
    } else {
      elements.searchLabel.classList.remove("hidden");
    }

    if (elements.input.value.length >= MIN_SEARCH_LENGTH) {
      const initial = index.search(elements.input.value);

      if (initial.length === 0) {
        const wildcardResults = index.search(`${elements.input.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${elements.input.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }
}

document.addEventListener("DOMContentLoaded", initializeSearch);