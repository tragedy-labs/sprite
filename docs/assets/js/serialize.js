const data = require("./searchData.js");
const lunr = require("./lunr.js");

const fs = require("fs");
const path = require("path");

const config = {
  minSearchLength: 3,
  indexedFields: ["name", "desc", "type", "parent"],
};

// Build the index using Lunr.js
const index = lunr(function () {
  this.ref("name");

  config.indexedFields.forEach((field) => {
    this.field(field);
  }, this);

  data.forEach((doc) => {
    this.add(doc);
  }, this);
});

fs.writeFileSync(
  path.join(__dirname, "./serialized.json"),
  JSON.stringify(index, null, 2)
);
