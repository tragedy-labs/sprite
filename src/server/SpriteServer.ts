import { ServerSession } from '../session/ServerSession.js';
import { ISpriteBaseSession } from '../session/Session.js';
import { SpriteDatabase } from '../database/SpriteDatabase.js';
import {
  ArcadeServerInformation,
  ArcadeServerInformationLevel,
  ISpriteCreateArcadeUser,
  Server,
  SpriteArcadeServerEvents
} from './Server.js';

/**
 * Methods for interact with an ArcadeDB server. Manage databases, users, etc.
 * @param {ISpriteRestClientConnectionParameters} parameters Connection details to access the server with.
 * @example
 *
 * const client = new SpriteServer({
 *   username: 'aUser',
 *   password: 'aPassword',
 *   address: 'http://localhost:2480',
 * });
 *
 * async function serverReadyExample() {
 *   try {
 *     const ready = await client.serverReady();
 *     if (ready) {
 *       console.log(ready);
 *       // true;
 *     }
 *   } catch (error) {
 *     throw new Error(
 *       'An error occurred while running example.',
 *       { cause: error }
 *     );
 *   }
 * }
 *
 * serverReadyExample();
 */
class SpriteServer {
  private _session: ServerSession;
  constructor(parameters: ServerSession);
  constructor(parameters: ISpriteBaseSession);
  constructor(parameters: ISpriteBaseSession | ServerSession) {
    if (parameters instanceof ServerSession) {
      this._session = parameters;
    } else {
      this._session = new ServerSession(parameters);
    }
  }
  /**
   * Returns a `boolean` value indicating if the ArcadeDB server is ready.\
   * Useful for remote monitoring of server readiness.
   * @returns `true` if the server is ready, otherwise `false`.
   * @example
   *
   * const client = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function serverReadyExample() {
   *   try {
   *     const ready = await client.serverReady();
   *     if (ready) {
   *       console.log(ready);
   *       // true;
   *     }
   *   } catch (error) {
   *     throw new Error(
   *       'An error occurred while running example.',
   *       { cause: error }
   *     );
   *   }
   * }
   *
   * serverReadyExample();
   */
  serverReady = async (): Promise<boolean> => Server.serverReady(this._session);
  /**
   * When you close a database in ArcadeDB, it:
   * 1. Frees up resources on the server: The database instance is released, and the associated resources, such as memory, threads, and file handles, are returned to the system. This helps to reduce the server's memory footprint and free up resources for other tasks.
   * 2. Releases it from RAM: The database instance is removed from the server's RAM, which means that the database's metadata, schema, and cached data are no longer stored in memory. This helps to reduce memory usage.
   * 3. Prevents further operations: Once the database is closed, users can no longer perform operations on the database, such as executing queries, creating new records, or modifying existing data. The database is effectively "offline" until it's reopened.
   *
   * @param databaseName The name of the database to close.
   * @returns The response from the server.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function closeDatabaseExample(databaseName: string) {
   *   try {
   *     const closed = await server.closeDatabase(databaseName);
   *     console.log(closed);
   *     // true
   *   } catch (error) {
   *     // manage error conditions
   *     console.error(error);
   *   }
   * };
   *
   * closeDatabaseExample('aDatabase');
   */
  closeDatabase = async (databaseName: string): Promise<boolean> =>
    Server.closeDatabase(this._session, databaseName);
  /**
   *
   * When you "open" a database in ArcadeDB, it means you're creating an instance of the database in memory, allocating resources, and making the database available for operations. Here's what happens when you open a database:
   *
   * 1. Create a new database instance: A new database instance is created in memory, which includes the database's metadata, schema, and cached data. This instance is used to manage the database's resources and provide access to the data.
   * 2. Allocate resources: The database instance allocates the necessary resources, such as memory, threads, and file handles, to support the database's operations. This ensures that the database has the necessary resources to handle incoming requests.
   * 3. Load database metadata and schema: The database's metadata and schema are loaded into memory, which includes information about the database's structure, indexes, and relationships.
   * 4. Connect to the underlying storage: The database instance establishes a connection to the underlying storage, such as disk storage, to access the database files.
   * 5. Make the database available for operations: The database is now available for users to perform operations, such as executing queries, creating new records, or modifying existing data.
   *
   * @param databaseName The name of the database to open.
   * @returns The response from the server.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function openDatabaseExample(databaseName: string) {
   *   try {
   *     const open = await server.openDatabase(databaseName);
   *     console.log(open);
   *     // true
   *   } catch (error) {
   *     // handle errors
   *     console.error(error);
   *   }
   * };
   *
   * openDatabaseExample('aDatabase');
   */
  openDatabase = async (databaseName: string): Promise<boolean> =>
    Server.openDatabase(this._session, databaseName);
  /**
   * Returns a `SpriteDatabase` instance for the supplied `databaseName`,
   * using the authorization details of the `SpriteServer` instance.
   * @param databaseName The name of the database to create a client for.
   * @returns An instance of `SpriteDatabase`.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function databaseExample() {
   *   try {
   *     const database = await server.database('aDatabase');
   *     // returns an instance of SpriteDatabase
   *     console.log(database.name);
   *   } catch (error) {
   *     // handle errors
   *     console.error(error);
   *   }
   * };
   *
   * databaseExample();
   */
  database = (databaseName: string): SpriteDatabase =>
    Server.database(this._session, databaseName);
  /**
   * Sends a command to the ArcadeDB server and returns the response.
   *
   * This method provides a way to execute arbitrary commands on the server, such as creating databases, executing queries, or performing administrative tasks.
   *
   * @param command The command to send to the server, such as `CREATE DATABASE aDatabase`. The command string should be a valid ArcadeDB command, and it's case-sensitive.
   * @returns The response from the server, simplified from the raw JSON response. The response object will contain the `result` property, which can be a boolean value (e.g., `OK`), a string, or an object. Other properties may include `user`, `version`, and `serverName`.
   * @throws If the command fails to execute, an error will be thrown with a message describing the problem. The error object will contain a `cause` property with the underlying error.
   *
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function commandExample(databaseName: string) {
   *   try {
   *     const response = await server.command(`CREATE DATABASE ${databaseName}`);
   *     console.log(response);
   *     // {
   *     //   user: 'aUser',
   *     //   version: '24.x.x',
   *     //   serverName: 'ArcadeDB_0',
   *     //   result: 'ok'
   *     // }
   *   } catch (error) {
   *     // Will throw an error for conditions such as:
   *     // Invalid credentials, Database Already Exists, etc.
   *     console.error(error);
   *     // {
   *     //   message: 'Encountered an error when sending a command to the server.',
   *     //   cause: Error: Invalid credentials
   *     // }
   *   }
   * }
   *
   * commandExample('aDatabase');
   */
  command = async <T>(command: string): Promise<T> =>
    Server.command(this._session, command);
  /**
   * Connects this server to a cluster with `address`.
   * @param address The address of the cluster to connect (i.e. 192.168.0.1)
   * @returns The response from the server.
   * @throws `Error` if the cluster could not be connected.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function connectClusterExample(address: string) {
   *   try {
   *     const connected = await server.connectCluster(address);
   *     console.log(connected);
   *     // true
   *   } catch (error) {
   *     console.log(error);
   *     // handle error conditions
   *   }
   * }
   *
   * connectClusterExample('192.168.0.1');
   */
  connectCluster = async (address: string): Promise<boolean> =>
    Server.connectCluster(this._session, address);
  /**
   * Create a database
   * @param databaseName The name of the database to create.
   * @returns An instance of `SpriteDatabase`, targeting the created database.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function createDatabaseExample(databaseName: string) {
   *   try {
   *     const database = await server.createDatabase(databaseName);
   *     console.log(database.name);
   *     // 'aDatabase'
   *   } catch (error) {
   *     console.log(error);
   *     // handle error conditions
   *   }
   * }
   *
   * createDatabaseExample('aDatabase');
   */
  createDatabase = async (databaseName: string) =>
    Server.createDatabase(this._session, databaseName);
  /**
   * Create a user. `username`, `password`, and access controls to multiple databases
   * can be established using the `databases` property of the input parameters.
   * The `databases` object uses 'groups' to grant access controls. Assigning
   * a user to groups within a specific database grants them the permissions associated
   * with those groups for a particular database.
   * @param username The `username` of the user to create.
   * @param password The `password` of the user to create.
   * @param databases An object of databases to add the user to, and their permissions (groups they belong to).
   * @returns `true` if the user was created successfully.
   * @throws `Error` if the user could not be created.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function createUserExample(details: ISpriteCreateArcadeUser) {
   *   try {
   *     const created = await server.createUser(details);
   *     console.log(created);
   *     // true
   *   } catch (error) {
   *     console.error(error);
   *     // handle error conditions
   *   }
   * }
   *
   * createUserExample({
   *   username: 'myUsername',
   *   password: 'myPassword',
   *   databases: {
   *     "FirstDatabase": "admin"
   *     "SecondDatabase": "user"
   *   }
   * });
   */
  createUser = async (params: ISpriteCreateArcadeUser): Promise<boolean> =>
    Server.createUser(this._session, params);
  /**
   * Disconnects the server from the cluster.
   * @returns The response from the server.
   * @throws `Error` if the cluster could not be disconnected.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function disconnectClusterExample() {
   *   try {
   *     const disconnected = await server.disconnectCluster();
   *     console.log(disconnected);
   *     // true
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * disconnectClusterExample();
   */
  disconnectCluster = async (): Promise<boolean> =>
    Server.disconnectCluster(this._session);
  /**
   * Drop a database
   * @param databaseName The name of the database to drop.
   * @returns `true` if successfully dropped.
   * @throws `Error` if the database could not be dropped.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function dropDatabaseExample(databaseName: string) {
   *   try {
   *     const dropped = await server.dropDatabase(databaseName);
   *     console.log(dropped);
   *     // true
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * dropDatabaseExample('aDatabase');
   */
  dropDatabase = async (databaseName: string): Promise<boolean> =>
    Server.dropDatabase(this._session, databaseName);
  /**
   * Drop a user from the ArcadeDB server.
   * @param username The `username` of the user to drop from the ArcadeDB server.
   * @returns `true` if the user was successfully dropped.
   * @throws `Error` if the user could not be dropped.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function dropUserExample(username: string) {
   *   try {
   *     const dropped = await server.dropUser(username);
   *     console.log(dropped);
   *     // true
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * dropUserExample('aUser');
   */
  dropUser = async (username: string): Promise<boolean> =>
    Server.dropUser(this._session, username);
  /**
   * Retrieves a list of server events, optionally a filename of the form
   * `server-event-log-yyyymmdd-HHMMSS.INDEX.jsonl` (where INDEX is a integer, i.e. 0)
   * can be given to retrieve older event logs.
   * @returns An object containing he server events from the server, and filenames of the associated logs.
   * @throws `Error` if there was a problem fetching the event logs.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function getServerEventsExample() {
   *   try {
   *     const events = await server.getServerEvents();
   *     console.log(events);
   *     // {
   *     //    events: [
   *     //      {
   *     //        "time":"2024-01-30 06:37:20.325",
   *     //        "type":"INFO","component":"Server",
   *     //        "message":"ArcadeDB Server started in \\u0027development\\u0027 mode (CPUs\\u003d8 MAXRAM\\u003d3.84GB)"}]
   *     //      }
   *     //    ],
   *     //    files: [
   *     //      "server-event-log-20240205-060106.13.jsonl",
   *     //      "server-event-log-20240204-185020.12.jsonl",
   *     //      "server-event-log-20240204-063235.11.jsonl",
   *     //      "server-event-log-20240131-205254.10.jsonl",
   *     //      "server-event-log-20240130-133802.9.jsonl"
   *     //    ]
   *     // }
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * getServerEventsExample();
   */
  getEvents = async (): Promise<SpriteArcadeServerEvents> =>
    Server.getEvents(this._session);
  /**
   * Returns the current configuration.
   * @param mode The level of informatio detail to return.
   * * `basic` returns minimal server information
   * * `default` returns full server configuration (default value when no parameter is given)
   * * `cluster` returns the cluster layout
   * @returns The response from the server.
   * @throws `Error` if the information could not be retrieved.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function getInformationExample(mode: ArcadeServerInformationLevel) {
   *   try {
   *     const information = await sprite.getInformation(mode);
   *     console.log(information);
   *     // {
   *     //   user: 'aUser',
   *     //   version: '24.x.x',
   *     //   serverName: 'ArcadeDB_0'
   *     // }
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * getInformationExample('basic');
   */
  getInformation = async <M extends ArcadeServerInformationLevel = 'default'>(
    mode?: M
  ): Promise<ArcadeServerInformation<M>> =>
    Server.getInformation(this._session, mode);
  /**
   * Returns a list of database names that are present on the server.
   * @returns {Promise<Array<string>>} A list (array) of database names present on the server.
   * @throws `Error` if the database list could not be retrieved.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function listDatabasesExample() {
   *   try {
   *     const list = await sprite.listDatabases();
   *     console.log(list);
   *     // [ 'databaseName' ]
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * listDatabasesExample();
   */
  listDatabases = async (): Promise<Array<string>> =>
    Server.listDatabases(this._session);
  /**
   * Gracefully shutdown the server.\
   * `TODO:` This works, in that it does shutdown the server, but the fetch throws
   * before it resolves, guessing because the server is shutting down. A CURL, however,
   * returns an empty `204` response as the documentation indicates.
   * @returns `true` if the server is successfully shutdown.
   * @throws `Error` if there is a problem attempting the shutdown.
   * @example
   *
   * const server = new SpriteServer({
   *   username: 'aUser',
   *   password: 'aPassword',
   *   address: 'http://localhost:2480',
   * });
   *
   * async function shutdownExample() {
   *   try {
   *     const shutdown = await server.shutdown();
   *     console.log(shutdown);
   *   } catch (error) {
   *     console.error(error);
   *     // handle error condition
   *   }
   * }
   *
   * shutdownExample();
   */
  shutdown = async (): Promise<boolean> => Server.shutdown(this._session);
}

export { SpriteServer };
