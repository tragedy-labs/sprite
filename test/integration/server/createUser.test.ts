// Lib
import { SpriteServer } from '@/server/SpriteServer.js';
import { variables } from '@test/variables.js';

// Testing
import { testClient } from './testClient.js';

const config = {
  username: 'createUsername',
  password: variables.password,
  databases: {
    [variables.databaseName]: '*'
  }
};

let newUserClient: SpriteServer;

describe('SpriteServer.createUser', () => {
  beforeAll(async () => {
    newUserClient = new SpriteServer({
      username: config.username,
      password: variables.password,
      address: 'http://localhost:2480'
    });

    // Ensure the user does not exist before starting the tests
    await testClient.dropUser(config.username).catch(() => {});
  });

  it('should create a user and return true', async () => {
    /* Arrange & Act */
    const created = await testClient.createUser(config);

    /* Assert */
    expect(created).toBe(true);
  });

  it('should allow new user to check status of database', async () => {
    /* Arrange & Act */
    const isReady = await newUserClient.serverReady();

    /* Assert */
    expect(isReady).toBe(true);
  });

  it('should propagate errors from the database', async () => {
    /* Arrange & Act */
    const createPromise = testClient.createUser(config);

    /* Assert */
    // Expected to reject because the user already exists
    await expect(createPromise).rejects.toMatchSnapshot();
  });

  it('should prevent dropped user from checking status', async () => {
    /* Arrange */
    const dropped = await testClient.dropUser(config.username);

    /* Act */
    const nonExistantUserPromise = newUserClient.serverReady();

    /* Assert */
    expect(dropped).toBe(true);
    await expect(nonExistantUserPromise).rejects.toMatchSnapshot();
  });
});
