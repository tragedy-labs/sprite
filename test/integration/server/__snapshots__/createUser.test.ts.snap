// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteServer.createUser should prevent dropped user from checking status 1`] = `[Error: Unable to check the server status.]`;

exports[`SpriteServer.createUser should propagate errors from the database 1`] = `[Error: Could not create user createUsername. In database(s): Imported]`;
